llm/payload.go

```diff
@@ -0,0 +1,230 @@
+package llm
+
+import (
+	"compress/gzip"
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"log/slog"
+	"os"
+	"path/filepath"
+	"runtime"
+	"strings"
+
+	"golang.org/x/exp/slices"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/jmorganca/ollama/gpu"
+)
+
+var errPayloadMissing = fmt.Errorf("expected payloads not included in this build of ollama")
+
+// TODO: error check this or initialize it in init()
+// TODO: this should be a determinate location to avoid required files disappearing
+var workDir, _ = os.MkdirTemp("", "ollama")
+
+func Init() error {
+	slog.Info("extracting embedded files", "dir", workDir)
+
+	// also extract the .metal shader on darwin/arm64
+	if runtime.GOOS == "darwin" && runtime.GOARCH == "arm64" {
+		err := extractFiles(workDir, "build/*/*/*/bin/ggml-metal.metal.gz")
+		if err != nil {
+			return fmt.Errorf("extract .metal shader: %v", err)
+		}
+
+		os.Setenv("GGML_METAL_PATH_RESOURCES", filepath.Join(workDir, "metal"))
+	}
+
+	var binGlob string
+	if runtime.GOOS == "windows" {
+		binGlob = "build/windows/*/*/bin/*/*"
+	} else {
+		binGlob = "build/*/*/*/bin/*"
+	}
+
+	// extract server libraries
+	err := extractFiles(workDir, binGlob)
+	if err != nil {
+		return fmt.Errorf("extract binaries: %v", err)
+	}
+
+	// TODO: print available variants
+	// slog.Info(fmt.Sprintf("Dynamic LLM libraries %v", variants))
+	// slog.Debug("Override detection logic by setting OLLAMA_LLM_LIBRARY")
+	var variants []string
+	for v := range available() {
+		variants = append(variants, v)
+	}
+	slog.Info(fmt.Sprintf("Dynamic LLM libraries %v", variants))
+	slog.Debug("Override detection logic by setting OLLAMA_LLM_LIBRARY")
+
+	return nil
+}
+
+// binary names may contain an optional variant separated by '_'
+// For example, "ollama_rocm_v6" and "ollama_rocm_v5" or "ollama_cpu" and "ollama_cpu_avx2"
+// Any library without a variant is the lowest common denominator
+func available() map[string]string {
+	// glob workDir for files that start with ollama_
+	pattern := filepath.Join(workDir, "*")
+
+	files, err := filepath.Glob(pattern)
```

Comment:
I think we'll need to reshuffle things a bit for rocm otherwise the symlink to the rocm lib dir might wind up being mistaken for a variant.